dl1_data_handler/reader.py

```diff
@@ -0,0 +1,408 @@
+                        }
+                    ])
+                for col_name in array_info:
+                    col = f.root.Array_Information._f_col(col_name)
+                    self.unprocessed_example_description.append(
+                        {
+                            'name': tel_type + '_' + col_name,
+                            'tel_type': tel_type,
+                            'base_name': col_name,
+                            'shape': (num_tels,) + col.shape,
+                            'dtype': col.dtype
+                            }
+                        )
+        # Add event info to description
+        for col in self.event_info:
+            self.unprocessed_example_description.append(
+                {
+                    'name': col,
+                    'tel_type': None,
+                    'base_name': None,
+                    'shape': f.root.Events._f_col(col).shape,
+                    'dtype': f.root.Events._f_col(col).dtype
+                    }
+                )
+
+        self.processor = DL1DataProcessor(
+            self.mode,
+            self.unprocessed_example_description,
+            transforms
+            )
+
+        # Definition of preprocessed example
+        self.example_description = self.processor.output_description
+
+    def _select_event_intensity(self, row, intensity_selection):
+        if intensity_selection is None or self.mode == 'mono':
+            return True
+        elif self.mode in ['stereo', 'multi-stereo']:
+            # TODO: define event-wise intensity
+            return True
+
+    @staticmethod
+    def _select_image_intensity(image_charge, intensity_selection):
+        intensity = np.sum(image_charge)
+        lower = intensity_selection.get('lower', -np.inf)
+        upper = intensity_selection.get('upper', np.inf)
+        return (lower < intensity < upper)
+
+    # Get a single telescope image from a particular event, uniquely
+    # identified by the filename, tel_type, and image table index.
+    # First extract a raw 1D vector and transform it into a 2D image using a
+    # mapping table.
+    def _get_image(self, filename, tel_type, image_index):
+
+        f = self.files[filename]
+        record = f.root._f_get_child(tel_type)[image_index]
+        query = "type == '{}'".format(tel_type)
+        length = [x['num_pixels'] for x
+                  in f.root.Telescope_Type_Information.where(query)][0]
+        num_channels = len(self.image_channels)
+        vector = np.empty(shape=(length + 1, num_channels), dtype=np.float32)
+        # An "empty" pixel at index 0 is used to fill blank areas in image
+        vector[0, :] = 0.0
+        # If the telescope didn't trigger, the image index is 0 and a blank
+        # image of all zeros with be loaded
+        for i, channel in enumerate(self.image_channels):
+            vector[1:, i] = record[channel]
+        image = self.image_mapper.map_image(vector, tel_type)
```

Comment:
Applying the image mapper in the _get_image function, and thus in the __getitem__ function might slow down the training. Ideally, it would be applied once, when creating the dataset. But it that case, the data need to be stored in RAM and this may not be possible when dealing with the whole array. What do you think ?

Reply:
That's a fair point, and yes, the idea here is to avoid having to load the entire dataset in RAM (or store a preprocessed dataset on disk). To minimize the effect of image mapping on training speed, @TjarkMiener (re)implemented all the CTLearn image mapping methods to operate as matrix transformations, optimizing performance at mapping time. The mapping matrices for the method(s) used are calculated at initialization. That way, the expensive computations (the slowest being cubic interpolation at ~30s for all telescope types) are only done once, and the map_image function is relatively fast.